File: src/audio/playback/state/VolumeProperty.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package audio.playback.state;

import javafx.beans.property.SimpleDoubleProperty;

import static java.lang.Math.abs;
import static util.Util.clip;

/** 
 * Double property for volume, with additional methods.  Makes sure the value 
 * is always within valid range.
 */
public final class VolumeProperty extends SimpleDoubleProperty {
    
    public static final double MIN = 0;
    public static final double MAX = 1;
    public static final double AVG = (MAX+MIN)/2;
    public static final double STEP = abs(MAX-MIN)/20;
    
    /** Initializes default value: 0.5 */
    public VolumeProperty() {
        super(AVG);
    }    
    
    /** Initializes to provided value. */
    public VolumeProperty(double v) {
        super(clip(MIN,v,MAX));
    }

    /** Sets the value. Value outside of minimal-maximal value range will be clipped. */
    @Override
    public void set(double v) {
        super.set(clip(MIN,v,MAX));
    }
    
    /** @return minimum volume value: 0 */
    public double getMin() {
        return MIN;
    }
    
    /** @return average volume value: 0.5 */
    public double getAverage() {
        return AVG;
    }
    
    /** @return maximum volume value: 1 */
    public double getMax() {
        return MAX;
    }
    
    /** Increment value by step. */
    public void inc() {
        set(get()+STEP);
    }

    /** Decrement value by step. */
    public void dec() {
        set(get()-STEP);
    }
    
}

File: src/player/main/sp/it/pl/audio/playback/VolumeProperty.kt
package sp.it.pl.audio.playback

import javafx.beans.property.SimpleDoubleProperty
import sp.it.pl.audio.playback.VolumeProperty.Companion.MAX
import sp.it.pl.audio.playback.VolumeProperty.Companion.MIN
import sp.it.pl.main.Double01
import sp.it.util.math.clip

/** Double property for playback volume. Value is always within valid range [MIN]-[MAX]. */
class VolumeProperty: SimpleDoubleProperty {

   /** Minimum volume value: [MIN] */
   val min: Double01 get() = MIN
   /** Average volume value: [AVG] */
   val average: Double01 get() = AVG
   /** Maximum volume value: [MAX] */
   val max: Double01 get() = MAX

   constructor(v: Double01): super(v.clip(MIN, MAX))

   /** Sets the value. Value outside of minimal-maximal value range will be clipped. */
   override fun set(v: Double01) {
      super.set(v.clip(MIN, MAX))
   }

   /** Increment value by [STEP]. */
   fun incByStep() = set(get() + STEP)

   /** Decrement value by [STEP]. */
   fun decByStep() = set(get() - STEP)

   companion object {
      const val MIN = 0.0
      const val MAX = 1.0
      const val AVG = 0.5
      const val STEP = 0.1

      /** @return logarithmic 0-1 value from linear 0-1 value */
      @JvmStatic fun linToLog(v: Double01): Double = v*v

   }

}